contracts/token/oft/v2/precrime/OFTV2View.sol

```diff
@@ -0,0 +1,71 @@
+// SPDX-License-Identifier: MIT
+
+pragma solidity ^0.8.0;
+
+import "@openzeppelin/contracts/token/ERC20/IERC20.sol";
+import "./IOFTV2View.sol";
+import "../OFTCoreV2.sol";
+
+contract OFTV2View is IOFTV2View {
+    using BytesLib for bytes;
+
+    OFTCoreV2 immutable oft;
+    ILayerZeroEndpoint immutable endpoint;
+
+    /// @notice Local decimals to shared decimals rate
+    uint immutable ld2sdRate;
+
+    constructor(address _oft) {
+        oft = OFTCoreV2(_oft);
+        endpoint = OFTCoreV2(_oft).lzEndpoint();
+
+        (, bytes memory data) = ICommonOFT(_oft).token().staticcall(abi.encodeWithSignature("decimals()"));
+        uint8 decimals = abi.decode(data, (uint8));
+        uint8 sharedDecimals = OFTCoreV2(_oft).sharedDecimals();
+        ld2sdRate = 10**(decimals - sharedDecimals);
+    }
+
+    function lzReceive(uint16 _srcChainId, bytes32 _scrAddress, bytes memory _payload, uint _totalSupply) external view virtual returns (uint) {
+        require(_isPacketFromTrustedRemote(_srcChainId, _scrAddress), "OFTV2View: not trusted remote");
+        uint amount = _decodePayload(_payload);
+        return _totalSupply + amount;
+    }
+
+    function _decodePayload(bytes memory _payload) internal view returns (uint) {
+        uint64 amountSD = _payload.toUint64(33);
+        return amountSD * ld2sdRate;
+    }
+
+    function getInboundNonce(uint16 _srcChainId, bytes32 _scrAddress) external view virtual returns (uint64) {
+        require(_isPacketFromTrustedRemote(_srcChainId, _scrAddress), "OFTV2View: not trusted remote");
+        bytes memory path = oft.trustedRemoteLookup(_srcChainId);
+        return endpoint.getInboundNonce(_srcChainId, path);
+    }
+
+    function isProxy() external view virtual returns (bool) {
+        return false;
+    }
+
+    function getTotalSupply() external view virtual returns (uint) {
+        return IERC20(address(oft)).totalSupply();
+    }
+
+    function _isPacketFromTrustedRemote(uint16 _srcChainId, bytes32 _srcAddress) internal view returns (bool) {
+        bytes memory path = oft.trustedRemoteLookup(_srcChainId);
+        if (path.length > 20) {
```

Comment:
think we should change this check to path.length > 32, in case the trusted remote is aptos where length is 32 and it's not a path

Reply:
This is taken from Stargate :) 
`path.length > 20` is only needed to remove the local EVM address from the path
We have another check later to ensure that remote address  <= 32

Reply:
what if the trusted remote is not a path and a 32 bytes aptos address? then it will still be slicked in this case?

Reply:
This precrime is for `OFTV2`, which inherits LzApp. Using a 32 bytes address rather than a path for trustedRemote would lead to a failed state here

Reply:
then it would be better to revert if insufficient length, instead of slice only if length > 20

Reply:
Is it better to add the following check?
`require(path.length >= 40 && path.length <= 52, "OFTV2View: invalid path length length");`

Reply:
yup looks good

Reply:
should be path.length == 40 || path.length == 52. if its a new non evm chain, we always set it to bytes 32

Reply:
Done